From 90a092c5b690128dc7143efeec6c89acf65f09b8 Mon Sep 17 00:00:00 2001
From: Michael Jeanson <mjeanson@debian.org>
Date: Sun, 4 Oct 2026 11:04:15 +0800
Subject: [PATCH] DEBIAN: Add sh4 to the list of supported architectures

Signed-off-by: Jon Bernard <jbernard@debian.org>
Signed-off-by: Michael Jeanson <mjeanson@debian.org>

[ Mingcong Bai: Please note that the original patch, despite specifying
  sh4, uses generic __sh__ define as the conditional - build tested on
  SuperH-3 (sh3). ]

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 include/Makefile.am           |  2 ++
 include/urcu/arch.h           |  5 ++++
 include/urcu/arch/superh.h    | 49 +++++++++++++++++++++++++++++++++++
 include/urcu/uatomic.h        |  2 ++
 include/urcu/uatomic/superh.h | 48 ++++++++++++++++++++++++++++++++++
 5 files changed, 106 insertions(+)
 create mode 100644 include/urcu/arch/superh.h
 create mode 100644 include/urcu/uatomic/superh.h

diff --git a/include/Makefile.am b/include/Makefile.am
index 5f53f4c..81250bd 100644
--- a/include/Makefile.am
+++ b/include/Makefile.am
@@ -20,6 +20,7 @@ nobase_include_HEADERS = \
 	urcu/arch/riscv.h \
 	urcu/arch/s390.h \
 	urcu/arch/sparc64.h \
+	urcu/arch/superh.h \
 	urcu/arch/tile.h \
 	urcu/arch/x86.h \
 	urcu/assert.h \
@@ -82,6 +83,7 @@ nobase_include_HEADERS = \
 	urcu/uatomic/riscv.h \
 	urcu/uatomic/s390.h \
 	urcu/uatomic/sparc64.h \
+	urcu/uatomic/superh.h \
 	urcu/uatomic/tile.h \
 	urcu/uatomic/uassert.h \
 	urcu/uatomic/x86.h \
diff --git a/include/urcu/arch.h b/include/urcu/arch.h
index 5147d30..a82cfd4 100644
--- a/include/urcu/arch.h
+++ b/include/urcu/arch.h
@@ -157,6 +157,11 @@
 #define URCU_ARCH_LOONGARCH 1
 #include <urcu/arch/loongarch.h>
 
+#elif defined(__sh__)
+
+#define URCU_ARCH_SUPERH 1
+#include <urcu/arch/superh.h>
+
 #else
 #error "Cannot build: unrecognized architecture, see <urcu/arch.h>."
 #endif
diff --git a/include/urcu/arch/superh.h b/include/urcu/arch/superh.h
new file mode 100644
index 0000000..c79bd13
--- /dev/null
+++ b/include/urcu/arch/superh.h
@@ -0,0 +1,49 @@
+#ifndef _URCU_ARCH_SUPERH_H
+#define _URCU_ARCH_SUPERH_H
+
+/*
+ * arch/superh.h: definitions for the SuperH architecture
+ *
+ * Copyright (c) 2021 Michael Jeanson <mjeanson@efficios.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <urcu/compiler.h>
+#include <urcu/config.h>
+#include <urcu/syscall-compat.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdlib.h>
+#include <sys/time.h>
+
+/*
+ * On Linux, define the membarrier system call number if not yet available in
+ * the system headers.
+ */
+#if (defined(__linux__) && !defined(__NR_membarrier))
+#define __NR_membarrier		378
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#include <urcu/arch/generic.h>
+
+#endif /* _URCU_ARCH_SUPERH_H */
diff --git a/include/urcu/uatomic.h b/include/urcu/uatomic.h
index 1af3cd8..6e2353c 100644
--- a/include/urcu/uatomic.h
+++ b/include/urcu/uatomic.h
@@ -97,6 +97,8 @@ static inline int cmm_to_c11(int mo)
 #include <urcu/uatomic/riscv.h>
 #elif defined(URCU_ARCH_LOONGARCH)
 #include <urcu/uatomic/loongarch.h>
+#elif defined(URCU_ARCH_SUPERH)
+#include <urcu/uatomic/superh.h>
 #else
 #error "Cannot build: unrecognized architecture, see <urcu/arch.h>."
 #endif
diff --git a/include/urcu/uatomic/superh.h b/include/urcu/uatomic/superh.h
new file mode 100644
index 0000000..e9fc38a
--- /dev/null
+++ b/include/urcu/uatomic/superh.h
@@ -0,0 +1,48 @@
+/*
+ * Atomic exchange operations for the SuperH architecture. Let GCC do it.
+ *
+ * Copyright (c) 2021 Michael Jeanson <mjeanson@efficios.com>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#ifndef _URCU_ARCH_UATOMIC_SUPERH_H
+#define _URCU_ARCH_UATOMIC_SUPERH_H
+
+#include <urcu/compiler.h>
+#include <urcu/system.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define UATOMIC_HAS_ATOMIC_BYTE
+#define UATOMIC_HAS_ATOMIC_SHORT
+#define UATOMIC_HAS_ATOMIC_INT
+#if (CAA_BITS_PER_LONG == 64)
+#define UATOMIC_HAS_ATOMIC_LLONG
+#endif
+
+#ifdef __cplusplus
+}
+#endif
+
+#include <urcu/uatomic/generic.h>
+
+#endif /* _URCU_ARCH_UATOMIC_SUPERH_H */
-- 
2.56.0

